File: cid/scripts/cid-tool-redact.py
#!/usr/bin/env python3
"""PostToolUse redactor. Reads the hook JSON on stdin, inspects the tool's
primary text output via CID /inspect/v1 (through cid-common.sh, so auth/context
have one implementation), and on a REDACT verdict substitutes the redacted text
back into that same field — preserving the tool's output shape. Log-only rules
return ALLOW and are simply recorded. See docs/CLAUDE_CODE_BUILD_PLAN.md."""
import json
import os
import re
import subprocess
import sys

HERE = os.path.dirname(os.path.abspath(__file__))


def locate(resp):
    """Return (get, set) for the tool's primary text field, shape-preserving.
    A None setter means the whole response is the string to replace."""
    if isinstance(resp, str):
        return (lambda: resp, None)
    if isinstance(resp, dict):
        if isinstance(resp.get("stdout"), str):  # Bash
            return (lambda: resp["stdout"],
                    lambda t: resp.__setitem__("stdout", t))
        f = resp.get("file")
        if isinstance(f, dict) and isinstance(f.get("content"), str):  # Read
            return (lambda: f["content"],
                    lambda t: f.__setitem__("content", t))
        if isinstance(resp.get("content"), str):  # generic
            return (lambda: resp["content"],
                    lambda t: resp.__setitem__("content", t))
    return (None, None)


# Tokens that precede the real command rather than being it.
_CMD_PREFIXES = {"sudo", "env", "command", "nohup", "time", "exec", "cd", "then", "do"}
_CMD_OK = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._+-]{0,31}$")


def command_word(command):
    """The command a Bash line actually runs, e.g. `git` / `pytest` / `docker`.

    Naively taking the first whitespace token recorded shell noise as work:
    `TOK=$(cat f)` was logged as the command `TOK=$(cat`. Walk past assignments,
    substitutions and wrappers until something that looks like a program name
    appears, and give up rather than guess.
    """
    # `cd x && git status` should report git, not cd or x, so each segment of the
    # line is considered until one names a program.
    for segment in re.split(r"&&|\|\||;|\||\n", command or ""):
        for raw in segment.strip().split():
            tok = raw.strip("(){}!$\"'`")
            if re.match(r"^[A-Za-z_][A-Za-z0-9_]*=", tok):
                _, _, rest = tok.partition("=")
                # FOO=bar cmd -> the value is data, keep walking. FOO=$(cmd ...)
                # -> the substitution is the work being done.
                if not rest.startswith(("$(", "`", '"$(', "'$(")):
                    continue
                tok = rest.lstrip("(){}$\"'`")
            tok = tok.split("/")[-1]  # /usr/bin/python3 -> python3
            if not tok:
                continue
            if tok in _CMD_PREFIXES:
                # These wrap or precede the real command; `cd` also eats its
                # argument, so drop the rest of this segment.
                if tok in ("cd", "then", "do"):
                    break
                continue
            return tok if _CMD_OK.match(tok) else ""
    return ""


def work_context(data):
    """What this tool touched, as identifiers only.

    Read/Edit/Write carry a file_path; Grep carries a search path; Bash carries a
    command whose first word (git, npm, pytest) says what kind of work is going
    on. The rest of a Bash command line is dropped on purpose — flags and
    arguments routinely contain tokens, hosts and credentials, and none of that
    is needed to answer "what were they working on".
    """
    tool = str(data.get("tool_name") or "")
    ti = data.get("tool_input")
    path = ""
    if isinstance(ti, dict):
        if isinstance(ti.get("file_path"), str):
            path = ti["file_path"]
        elif isinstance(ti.get("path"), str):
            path = ti["path"]
        elif isinstance(ti.get("command"), str):
            path = command_word(ti["command"])
    return tool, path


def cid_inspect(text, session_id="", tool_name="", tool_path=""):
    """Inspect `text` via the shared shell helper; return the verdict JSON str.
    session_id comes from the hook payload (Claude Code does not export
    CLAUDE_SESSION_ID to hooks) and rides to the helper as CID_SESSION_ID so
    the ctx-file lookup and telemetry both use the real session."""
    env = dict(os.environ)
    if session_id:
        env["CID_SESSION_ID"] = session_id
    if tool_name:
        env["CID_TOOL_NAME"] = tool_name
    if tool_path:
        env["CID_TOOL_PATH"] = tool_path
    try:
        out = subprocess.run(
            ["sh", "-c",
             '. "$1/cid-common.sh"; cid_inspect response "$2" tool_use',
             "_", HERE, text],
            capture_output=True, text=True, env=env,
            timeout=int(os.environ.get("CID_INSPECT_TIMEOUT", "6")),
        )
        return out.stdout.strip()
    except Exception:
        return ""


def main():
    try:
        data = json.load(sys.stdin)
    except Exception:
        return
    resp = data.get("tool_response")
    get, setf = locate(resp)
    if get is None:
        return
    text = get()
    if not text or not text.strip():
        return

    session_id = str(data.get("session_id") or "")
    tool_name, tool_path = work_context(data)
    verdict_json = cid_inspect(text, session_id, tool_name, tool_path)
    if not verdict_json:
        return  # fail-open: tool already ran locally; leave output as-is
    try:
        v = json.loads(verdict_json)
    except Exception:
        return
    if v.get("verdict") != "REDACT" or not isinstance(v.get("redactedPayload"), str):
        return  # ALLOW (incl. flag/log-only) — nothing to substitute

    redacted = v["redactedPayload"]
    updated = redacted if setf is None else (setf(redacted) or resp)
    print(json.dumps({"hookSpecificOutput": {
        "hookEventName": "PostToolUse",
        "updatedToolOutput": updated,
        "additionalContext": "CID DLP redacted sensitive values in this tool "
                             "output before you received them.",
    }}))


if __name__ == "__main__":
    main()

File: cid/scripts/cid-tool-redact.js
"use strict";
// PostToolUse hook (Read|Grep|Bash|PowerShell) — the DLP redactor. Reads the
// hook JSON on stdin, inspects the tool's primary text output via CID
// /inspect/v1 (through cid-common.js, so auth/context have one implementation),
// and on a REDACT verdict substitutes the redacted text back into that same
// field — preserving the tool's output shape, which Claude Code requires for
// `updatedToolOutput` on built-in tools. Log-only rules return ALLOW and are
// simply recorded. See docs/CLAUDE_CODE_BUILD_PLAN.md.
//
// Fail-open: any error leaves the tool output untouched (the tool has already
// run locally by the time this hook fires).

const cid = require("./cid-common.js");

// Return {get, set} for the tool's primary text field, shape-preserving.
// A null setter means the whole response is the string to replace.
function locate(resp) {
  if (typeof resp === "string") {
    return { get: () => resp, set: null };
  }
  if (resp && typeof resp === "object" && !Array.isArray(resp)) {
    if (typeof resp.stdout === "string") {
      // Bash / PowerShell
      return { get: () => resp.stdout, set: (t) => { resp.stdout = t; } };
    }
    const f = resp.file;
    if (f && typeof f === "object" && typeof f.content === "string") {
      // Read
      return { get: () => f.content, set: (t) => { f.content = t; } };
    }
    if (typeof resp.content === "string") {
      // generic
      return { get: () => resp.content, set: (t) => { resp.content = t; } };
    }
  }
  return { get: null, set: null };
}

// Tokens that precede the real command rather than being it.
const CMD_PREFIXES = new Set(["sudo", "env", "command", "nohup", "time", "exec", "cd", "then", "do"]);
const CMD_OK = /^[A-Za-z0-9][A-Za-z0-9._+-]{0,31}$/;

// Python str.strip(chars) / lstrip(chars) equivalents.
function stripChars(s, chars) {
  let a = 0;
  let b = s.length;
  while (a < b && chars.includes(s[a])) a++;
  while (b > a && chars.includes(s[b - 1])) b--;
  return s.slice(a, b);
}
function lstripChars(s, chars) {
  let a = 0;
  while (a < s.length && chars.includes(s[a])) a++;
  return s.slice(a);
}

// The command a Bash line actually runs, e.g. `git` / `pytest` / `docker`.
//
// Naively taking the first whitespace token recorded shell noise as work:
// `TOK=$(cat f)` was logged as the command `TOK=$(cat`. Walk past assignments,
// substitutions and wrappers until something that looks like a program name
// appears, and give up rather than guess.
function commandWord(command) {
  // `cd x && git status` should report git, not cd or x, so each segment of the
  // line is considered until one names a program.
  for (const segment of String(command || "").split(/&&|\|\||;|\||\n/)) {
    const parts = segment.trim().split(/\s+/).filter(Boolean);
    let broke = false;
    for (const rawTok of parts) {
      let tok = stripChars(rawTok, "(){}!$\"'`");
      if (/^[A-Za-z_][A-Za-z0-9_]*=/.test(tok)) {
        const rest = tok.slice(tok.indexOf("=") + 1);
        // FOO=bar cmd -> the value is data, keep walking. FOO=$(cmd ...)
        // -> the substitution is the work being done.
        if (!(rest.startsWith("$(") || rest.startsWith("`") || rest.startsWith('"$(') || rest.startsWith("'$("))) {
          continue;
        }
        tok = lstripChars(rest, "(){}$\"'`");
      }
      const segs = tok.split("/");
      tok = segs[segs.length - 1]; // /usr/bin/python3 -> python3
      if (!tok) continue;
      if (CMD_PREFIXES.has(tok)) {
        // These wrap or precede the real command; `cd` also eats its
        // argument, so drop the rest of this segment.
        if (tok === "cd" || tok === "then" || tok === "do") {
          broke = true;
          break;
        }
        continue;
      }
      return CMD_OK.test(tok) ? tok : "";
    }
    if (broke) continue;
  }
  return "";
}

// What this tool touched, as identifiers only.
//
// Read/Edit/Write carry a file_path; Grep carries a search path; Bash carries a
// command whose first word (git, npm, pytest) says what kind of work is going
// on. The rest of a Bash command line is dropped on purpose — flags and
// arguments routinely contain tokens, hosts and credentials, and none of that
// is needed to answer "what were they working on".
function workContext(data) {
  const tool = String((data && data.tool_name) || "");
  const ti = data && data.tool_input;
  let p = "";
  if (ti && typeof ti === "object") {
    if (typeof ti.file_path === "string") p = ti.file_path;
    else if (typeof ti.path === "string") p = ti.path;
    else if (typeof ti.command === "string") p = commandWord(ti.command);
  }
  return { toolName: tool, toolPath: p };
}

async function main() {
  const data = cid.parseJson(cid.readStdin());
  if (!data) return;

  const resp = data.tool_response;
  const { get, set } = locate(resp);
  if (!get) return;
  const text = get();
  if (!text || !text.trim()) return;

  // session_id comes from the hook payload (Claude Code does not export
  // CLAUDE_SESSION_ID to hooks) so the ctx-file lookup and telemetry both use
  // the real session.
  process.env.CID_SESSION_ID = cid.sessionFromHook(data);
  const { toolName, toolPath } = workContext(data);

  const verdictJson = await cid.inspect("response", text, "tool_use", { toolName, toolPath });
  if (!verdictJson) return; // fail-open: tool already ran locally; leave output as-is

  const v = cid.parseJson(verdictJson);
  if (!v) return;
  if (v.verdict !== "REDACT" || typeof v.redactedPayload !== "string") {
    return; // ALLOW (incl. flag/log-only) — nothing to substitute
  }

  const redacted = v.redactedPayload;
  let updated;
  if (set === null) {
    updated = redacted;
  } else {
    set(redacted);
    updated = resp;
  }

  process.stdout.write(
    JSON.stringify({
      hookSpecificOutput: {
        hookEventName: "PostToolUse",
        updatedToolOutput: updated,
        additionalContext:
          "CID DLP redacted sensitive values in this tool output before you received them.",
      },
    }) + "\n"
  );
}

cid.run(main);
